numba/tests/test_function.py

```diff
@@ -0,0 +1,585 @@
+            return op(foo)(choose_left)
+
+        self.assertEqual(w(njit, a, b, True),
+                         w(lambda f:f, a.pyfunc, b.pyfunc, True))
+        self.assertEqual(w(njit, a, b, False),
+                         w(lambda f:f, a.pyfunc, b.pyfunc, False))
+        self.assertNotEqual(w(njit, a, b, True),
+                            w(lambda f:f, a.pyfunc, b.pyfunc, False))
+
+    def _test_in_choose_out(self, decor):
+
+        @decor
+        def a(i):
+            return i + 1
+
+        @decor
+        def b(i):
+            return i + 2
+
+        def w(op, a, b, choose_left):
+            def foo(a, b, choose_left):
+                if choose_left:
+                    return a
+                else:
+                    return b
+            return op(foo)(a, b, choose_left)
+
+        self.assertEqual(w(njit, a, b, True), w(lambda f:f, a, b, True))
+        self.assertEqual(w(njit, a, b, False), w(lambda f:f, a, b, False))
+        self.assertNotEqual(w(njit, a, b, True), w(lambda f:f, a, b, False))
+
+    def _test_in_choose_func(self, decor):
+
+        @decor
+        def a(i):
+            return i + 1
+
+        @decor
+        def b(i):
+            return i + 2
+
+        def w(op, a, b, choose_left):
+            def foo(a, b, choose_left):
+                if choose_left:
+                    f = a
+                else:
+                    f = b
+                return f(1)
+            return op(foo)(a, b, choose_left)
+
+        self.assertEqual(w(njit, a, b, True),
+                         w(lambda f:f, a.pyfunc, b.pyfunc, True))
+        self.assertEqual(w(njit, a, b, False),
+                         w(lambda f:f, a.pyfunc, b.pyfunc, False))
+        self.assertNotEqual(w(njit, a, b, True),
+                            w(lambda f:f, a.pyfunc, b.pyfunc, False))
+
+    def _test_in_pick_func_call(self, decor):
+
+        @decor
+        def a(i):
+            return i + 1
+
+        @decor
+        def b(i):
+            return i + 2
+
+        def w(op, a, b, index):
+            def foo(funcs, i):
+                r = funcs[i](123)
+                return r
+            return op(foo)((a, b), index)
+
+        self.assertEqual(w(njit, a, b, 0), w(lambda f:f, a.pyfunc, b.pyfunc, 0))
+        self.assertEqual(w(njit, a, b, 1), w(lambda f:f, a.pyfunc, b.pyfunc, 1))
+        self.assertNotEqual(w(njit, a, b, 0),
+                            w(lambda f:f, a.pyfunc, b.pyfunc, 1))
+
+    def _test_in_iter_func_call(self, decor):
+
+        @decor
+        def a(i):
+            return i + 1
+
+        @decor
+        def b(i):
+            return i + 2
+
+        def w(op, a, b):
+            def foo(funcs, n):
+                r = 0
+                for i in range(n):
+                    f = funcs[i]
+                    r = r + f(r)
+                return r
+            return op(foo)((a, b), 2)
+
+        self.assertEqual(w(njit, a, b), w(lambda f:f, a.pyfunc, b.pyfunc))
+
+    def test_all(self):
```

Comment:
The TestFunctionType class currently implements 17 usage patterns (methods starting with `_test`) of first-class functions that could be of 6 different kinds (see `all_func_kinds` list).
The `test_all` method applies the test methods to all different input types so that there will be 102 tests run in total. Although this PR implements full support for one kind only (`cfunc_func`), it seemed impractical and error-prone to have the same test pattern repeated possibly 6 times.

Ideally, the above should be implemented using parameterized test methods (e.g. [pytest.mark.parametrize](https://docs.pytest.org/en/latest/parametrize.html#parametrize)) but std unittest does not support this. There is [parameterized](https://github.com/wolever/parameterized) project for unittest though.

Ahh, I just found `https://docs.python.org/3/library/unittest.html#subtests`... and fix the above.

Reply:
Fixed. The tests are now explicit in what function types are supported and what not-yet supported.